File: protobuf-serialization/src/test/java/com/stannis/benchmark/mapper/AbstractMemoryOprationCaptureTemplate.java
package com.stannis.benchmark.mapper;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Attempt to determine how much memory is used during a test
 */
public class AbstractMemoryOprationCaptureTemplate implements MemoryTest {

    private static final Logger LOG = LoggerFactory.getLogger(AbstractMemoryOprationCaptureTemplate.class);

    private Runtime runtime;

    AbstractMemoryOprationCaptureTemplate() {
        this.runtime = Runtime.getRuntime();
    }

    protected void logMemoryUsage(long memory, long after, final String Message) {
        LOG.info(Message + " = {} bytes", bytesToMegabytes((after - memory)));
    }

    protected long getMemoryAfterOperation() {
        return runtime.totalMemory() - runtime.freeMemory();
    }

    protected long getInitialMemory() {
        runtime.gc();
        return runtime.totalMemory() - runtime.freeMemory();
    }

}

File: protobuf-bench/src/test/kotlin/com/stannis/protobufbench/mapper/AbstractMemoryOprationCaptureTemplate.kt
package com.stannis.protobufbench.mapper

import org.slf4j.LoggerFactory

/**
 * Attempt to determine how much memory is used during a test
 */
class AbstractMemoryOprationCaptureTemplate internal constructor() : MemoryTest {
    private val runtime: Runtime

    init {
        runtime = Runtime.getRuntime()
    }

    protected fun logMemoryUsage(memory: Long, after: Long, Message: String) {
        LOG.info("$Message = {} bytes", bytesToMegabytes(after - memory))
    }

    protected val memoryAfterOperation: Long
        protected get() = runtime.totalMemory() - runtime.freeMemory()
    protected val initialMemory: Long
        protected get() {
            runtime.gc()
            return runtime.totalMemory() - runtime.freeMemory()
        }

    companion object {
        private val LOG = LoggerFactory.getLogger(AbstractMemoryOprationCaptureTemplate::class.java)
    }
}

